### PR TITLE
Add validation for producer methods/fields and add automated test.

### LINKED ISSUE #4177: Validate unproxyable bean types of normal scoped producers
https://docs.jboss.org/cdi/spec/2.0/cdi-spec.html#unproxyable

### PR DESCRIPTION
**Pull request by manovotn**
Fixes #4177

### CODE REVIEW

**Comment by manovotn** [comment]
Adds proxyability checks for producers, currently checks return type and then constructor of that class - all of it on assumption that the producer is for normal-scoped bean and that it is not an interface type.

File: independent-projects/arc/processor/src/main/java/io/quarkus/arc/processor/BeanInfo.java [hunk 1]
```diff
@@ -364,6 +365,61 @@ void validate(List<Throwable> errors, List<BeanDeploymentValidator> validators)
                                         "%s bean is not proxyable because it has a private no-args constructor: %s. To fix this problem, change the constructor to be package-private",
                                         classifier, this)));
             }
+        } else if (isProducerMethod()) {
+            MethodInfo methodInfo = target.get().asMethod();
```
Comment (mkouba) [inline independent-projects/arc/processor/src/main/java/io/quarkus/arc/processor/BeanInfo.java]: I believe that the code for producer methods and fields could be shared?

File: independent-projects/arc/tests/src/test/java/io/quarkus/arc/test/validation/AbstractNormalScopedFinalTest.java [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+
+        @Produces
+        @ApplicationScoped
+        public FinalFoo createFoo() {
+            return new FinalFoo();
+        }
+    }
+
+    @ApplicationScoped
+    static class FieldProducerWithFinalWrongConstructor {
```
Comment (jmartisk) [inline independent-projects/arc/tests/src/test/java/io/quarkus/arc/test/validation/AbstractNormalScopedFinalTest.java]: Why the word `Final` here? I don't see anything final in this case

Comment (mkouba) [reply independent-projects/arc/tests/src/test/java/io/quarkus/arc/test/validation/AbstractNormalScopedFinalTest.java]: Good catch ;-). It should be probably `FieldProducerWithWrongConstructor`.

File: independent-projects/arc/processor/src/main/java/io/quarkus/arc/processor/BeanInfo.java (hunk 1 above)
Comment (manovotn) [reply independent-projects/arc/processor/src/main/java/io/quarkus/arc/processor/BeanInfo.java]: Good point

**Review by mkouba — APPROVED** [review]

**Pull request merged**